### PR TITLE
refactor!: embed classes

### PR DESCRIPTION
**Pull request by Om1609**
## Summary

Attributes for Author, Footer, Image, Thumbnail, Video, and Provider on the embed will now return `None` instead of `EmbedProxy`, when they are not set. This change is done for correct API reflection.

This allowed the complete removal of `EmbedProxy` in favour of separate classes for the above attributes. This allows for concrete typing support. The new classes are `EmbedAuthor`, `EmbedFooter`, `EmbedMedia` and `EmbedProxy`

`Embed.Empty` and `EmptyEmbed` have now been removed in favour of `None`

## Information

- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [x] If `type: ignore` comments were used, a comment is also left explaining why.
- [x] I have updated the changelog to include these changes.

### CODE REVIEW

File: discord/embeds.py [hunk 1]
```diff
@@ -123,16 +122,23 @@ class EmbedAuthor(EmbedProxy):
...
-        proxy_icon_url: MaybeEmpty[str] = EmptyEmbed,
+        url: str | None = None,
+        icon_url: str | None = None,
+        proxy_icon_url: str | None = None,
     ) -> None:
+        # layer = {
+        #     k: v
+        #     for k, v in locals().items()
+        #     if k in {"name", "url", "icon_url", "proxy_icon_url"} and v
+        # }
```
Comment (Lulalaby) [inline discord/embeds.py:133]: Same

**Review by Lulalaby — CHANGES_REQUESTED** [review]

**Comment by Lulalaby** [comment]
nice work om

File: discord/embeds.py [hunk 2]
```diff
@@ -123,20 +65,32 @@ class EmbedAuthor(EmbedProxy):
...
     ) -> None:
-        layer = {
-            k: v
-            for k, v in locals().items()
-            if k in {"name", "url", "icon_url", "proxy_icon_url"}
-            and v is not EmptyEmbed
+        self.name: str = name
+        self.url: str | None = url
+        self.icon_url: str | None = icon_url
+        self.proxy_icon_url: str | None = None
```
Comment (NeloBlivion) [inline discord/embeds.py:74]: The old `not` check was necessary to ensure `None` never ended up in the payload.
```py
discord.Embed(author=discord.EmbedAuthor(name="Nelo"))
```
Raises when sent
```py
discord.errors.HTTPException: 400 Bad Request (error code: 50035): Invalid Form Body
In embeds.0.author.icon_url: Not a well formed URL.
In embeds.0.author.url: Scheme "none" is not supported. Scheme must be one of ('http', 'https'). Not a well formed URL.
```
Please recreate the previous logic as you see fit.

File: discord/embeds.py [hunk 3]
```diff
@@ -123,20 +65,32 @@ class EmbedAuthor(EmbedProxy):
...
+        self.url = data.get("url")
+        self.icon_url = data.get("icon_url")
+        self.proxy_icon_url = data.get("proxy_icon_url")
+        return self
+
+    def to_dict(self) -> dict[str, str | None]:
+        return {
+            "name": str(self.name),
+            "url": str(self.url),
+            "icon_url": str(self.icon_url),
```
Comment (NeloBlivion) [inline discord/embeds.py:89]: The string conversion here is unnecessary and would also lead to a literal `"None"` being returned in urls.
```suggestion
            "name": self.name,
            "url": self.url,
            "icon_url": self.icon_url,
```

File: discord/embeds.py [hunk 4]
```diff
@@ -152,15 +106,53 @@ class EmbedFooter(EmbedProxy):
...
+        self = cls.__new__(cls)
+        self.text = data.get("text")
+        self.icon_url = data.get("icon_url")
+        self.proxy_icon_url = data.get("proxy_icon_url")
+        return self
+
+    def to_dict(self) -> dict[str, Any]:
+        return {
+            "text": str(self.text),
+            "icon_url": str(self.icon_url),
```
Comment (NeloBlivion) [inline discord/embeds.py:126]: Same as comment for `EmbedAuthor.to_dict`
```suggestion
            "text": self.text,
            "icon_url": self.icon_url,
```

File: discord/embeds.py [hunk 5]
```diff
@@ -152,15 +106,53 @@ class EmbedFooter(EmbedProxy):
...
-        proxy_icon_url: MaybeEmpty[str] = EmptyEmbed,
+        icon_url: str | None = None,
     ) -> None:
-        layer = {
-            k: v
-            for k, v in locals().items()
-            if k in {"text", "icon_url", "proxy_icon_url"} and v is not EmptyEmbed
+        self.text: str = text
+        self.icon_url: str | None = icon_url
+        self.proxy_icon_url: str | None = None
```
Comment (NeloBlivion) [inline discord/embeds.py:113]: Same as comment for `EmbedAuthor.__init__`; please reimplement checks for `None`.

**Review by NeloBlivion — CHANGES_REQUESTED** [review]
All in all a good effort, mainly needs some fine tuning.
Please update all necessary docs;
- attrs like `author`, `footer` and `video` should refer to their new class.
- Remove `Empty` and `EmbedProxy` references.
- Add **all** new classes such as `EmbedMedia` to [data_classes.rst](https://github.com/Om1609/pycord/blob/embed/docs/api/data_classes.rst?plain=1).

Any changes you make to docs can be previewed [here](https://pycord--2063.org.readthedocs.build/en/2063/api/data_classes.html).

File: discord/embeds.py (hunk 2 above)
Comment (Om1609) [reply discord/embeds.py:74]: Ah that's right. Working on it

File: discord/embeds.py (hunk 3 above)
Comment (Om1609) [reply discord/embeds.py:89]: Wouldn't removing the string conversion for `name` cause issues if anything else like an interger is passed?

Comment (Lulalaby) [reply discord/embeds.py]: ppl shouldn't submit ints. their fault

File: discord/embeds.py (hunk 4 above)
Comment (Lulalaby) [reply discord/embeds.py]: see https://github.com/Pycord-Development/pycord/pull/2063#discussion_r1190304216

File: discord/embeds.py (hunk 5 above)
Comment (Lulalaby) [reply discord/embeds.py:126]: see https://github.com/Pycord-Development/pycord/pull/2063#discussion_r1190304216

**Review by Lulalaby — DISMISSED** [review]

**Review by Lulalaby — APPROVED** [review]

**Review by NeloBlivion — APPROVED** [review]

**Comment by Revnoplex** [comment]
Not sure if this has been noted as part of the breaking changes but I noticed another breaking change which was the `fields` argument of `Embed` no longer accepted `None`
which would cause `'NoneType' object is not iterable` here:
```python
File "bot-dir/venv/lib/python3.11/site-packages/discord/embeds.py", line 995, in to_dict
    result["fields"] = [field.to_dict() for field in self._fields]
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by Om1609** [comment]
Yes that seems like a missed bug. I'll fix this in a bit

**Comment by Om1609** [comment]
> Yes that seems like a missed bug. I'll fix this in a bit

Already fixed by #2071

**Pull request merged**